Jakartaee auth0 (#1332)
oktadeveloper/okta-blog: _source/_posts/2023-03-01-jakarta-ee-oidc.md

    │   ├── logging.properties  // Simple console logging configuration
    │   └── oidc.properties  // OpenID Connect config properties
    └── webapp
        └── WEB-INF
            └── jboss-web.xml  // Configures context root to '/'
```

When the application loads, the OpenID Connect properties are loaded from `oidc.properties` by the `OidcConfig` class. These values are used by the `JwtFilter` to create the class that verifies JSON Web Tokens. These properties are also used by the `ProtectedServlet` in the `@OpenIdAuthenticationMechanismDefinition` annotation to configure OIDC.

The `jboss-web.xml` file is simply used to change the context root to `/`.

Neither of the files in the `META-INF` directory seem to be required for the application to function. The `beans.xml` file explicitly enables CDI / dependency injection. However, this would also be done implicitly by the use of the annotations. The `MANIFEST.MF` file defines some provided runtime dependencies related to CDI. Perhaps in some runtime environments it would be necessary to include this file, but it seems unneeded when running locally with WilfFly.

## Create an Auth0 OIDC application

> If you have not already, install the [Auth0 CLI](https://github.com/auth0/auth0-cli) and run `auth0 login` in a terminal. As I write this, the Auth0 CLI 1.0 version is in beta. It adds some new features that I'll mention. You can take a look at [the release here](https://github.com/auth0/auth0-cli/releases/tag/v1.0.0-beta.1).

Reviewer: ```suggestion
If you have not already, [install the Auth0 CLI](https://github.com/auth0/auth0-cli#installation) and run `auth0 login` in a terminal. As I write this, the Auth0 CLI 1.0 version is in beta. It adds some new features that I'll mention. You can take a look at [the release here](https://github.com/auth0/auth0-cli/releases/tag/v1.0.0-beta.1).
```

Change the code for the action to the following.

```js
exports.onExecutePostLogin = async (event, api) => {
  const namespace = 'http://www.jakartaee.demo';
  if (event.authorization) {
    api.idToken.setCustomClaim('preferred_username', event.user.email);
    api.idToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
    api.accessToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
  }
}
```

Click on **Deploy**. 

Click on the **Add to flow** link in the popup window that slides in (if you miss this, you can find the new action under the custom action tab back in the flow panel).

Drag the **Add Roles** action over under the **Rules (legacy)** action. 

{% img blog/jakartaee-auth0/auth0-action-flow.png alt:"Auth0 Action Flow" width:"400" %}{: .center-image }

Click **Apply** (top right of the panel).

## Explore the `ProtectedServlet` and OIDC flow

Let's look at the `ProtectedServlet` first. This is the class that defines the OIDC annotation and will redirect to Auth0 to handle OIDC authentication.

`src/main/java/com/demo/ProtectedServlet.java`

```java
package com.demo;

...

@OpenIdAuthenticationMechanismDefinition(

>     providerURI = "${openIdConfig.issuerUri}",
>     clientId = "${openIdConfig.clientId}",
>     clientSecret = "${openIdConfig.clientSecret}",
>     redirectURI = "${baseURL}/callback",
>     // default 500ms caused timeouts for me
>     jwksConnectTimeout = 5000,
>     jwksReadTimeout = 5000,
>     extraParameters = {"audience=https://${openIdConfig.issuerUri}/api/v2/"},
>     claimsDefinition = @ClaimsDefinition(callerGroupsClaim = "http://www.jakartaee.demo/roles")

Reviewer: ```suggestion
    providerURI = "${oidcConfig.issuerUri}",
    clientId = "${oidcConfig.clientId}",
    clientSecret = "${oidcConfig.clientSecret}",
    redirectURI = "${baseURL}/callback",
    // default 500ms caused timeouts for me
    jwksConnectTimeout = 5000,
    jwksReadTimeout = 5000,
    extraParameters = {"audience=https://${oidcConfig.issuerUri}/api/v2/"},
    claimsDefinition = @ClaimsDefinition(callerGroupsClaim = "http://www.jakartaee.demo/roles")
```

When a user that is not authenticated attempts to load this resource, they are redirected to Auth0 for authentication. From a browser, the user sees Auth0's login screen. After successfully logging in, the user is redirected back to the `/callback` servlet with an authentication code. Jakarta EE's security framework intercepts this redirect and sends the code back to Auth0 to exchange it for an authentication token before passing control back to the `/callback` endpoint. 

At this point, the user is successfully authenticated. If you look at the callback servlet (shown below), you'll see that it simply redirects the user back to the `/protected` servlet. 

```java
package com.demo;

...

@WebServlet("/callback")
public class CallbackServlet extends HttpServlet {

    private static final Logger LOGGER = Logger.getLogger(CallbackServlet.class.getName());

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)

>             throws ServletException, IOException {

Reviewer: ```suggestion
        throws ServletException, IOException {
```

## Use a JWT to access the protected API

The `ApiServlet` file defines an API servlet.

`src/main/java/com/demo/ApiServlet.java`

```java
package com.demo;

...

@WebServlet("/api/protected")
public class ApiServlet extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {

        DecodedJWT jwt = (DecodedJWT) request.getAttribute("jwt");
        response.setContentType("text/plain");
        response.getWriter().println("Welcome, " + jwt.getClaims().get("sub"));
        response.getWriter().println(jwt.getClaims());
    }
}

This servlet by itself is not at all secure and would be public without the `JwtFilter` class, which is shown below. The filter intercepts any requests matching the `/api/*` URL pattern and denies them if they do not have a valid JWT.

`src/main/java/com/demo/JwtFilter.java`

```java
package com.demo;

...

@WebFilter(filterName = "jwtFilter", urlPatterns = "/api/*")
public class JwtFilter implements Filter {

    private static final Logger LOGGER = Logger.getLogger(JwtFilter.class.getName());

    @Inject

>     OidcConfig oidcConfig;
>     @Override

Reviewer: ```suggestion
    OidcConfig oidcConfig;

    @Override
```

    public void init(FilterConfig filterConfig) {

>         LOGGER.info("Auth0 jwtVerifier initialized for issuer:" + openIdConfig.getIssuerUri());

Reviewer: ```suggestion
        LOGGER.info("Auth0 jwtVerifier initialized for issuer:" + oidcConfig.getIssuerUri());
```

## Use a JWT to access the protected API

The `ApiServlet` file defines an API servlet.

`src/main/java/com/demo/ApiServlet.java`

```java
package com.demo;

...

@WebServlet("/api/protected")
public class ApiServlet extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {

        DecodedJWT jwt = (DecodedJWT) request.getAttribute("jwt");
        response.setContentType("text/plain");
        response.getWriter().println("Welcome, " + jwt.getClaims().get("sub"));
        response.getWriter().println(jwt.getClaims());
    }
}

This servlet by itself is not at all secure and would be public without the `JwtFilter` class, which is shown below. The filter intercepts any requests matching the `/api/*` URL pattern and denies them if they do not have a valid JWT.

`src/main/java/com/demo/JwtFilter.java`

```java
package com.demo;

...

@WebFilter(filterName = "jwtFilter", urlPatterns = "/api/*")
public class JwtFilter implements Filter {

    private static final Logger LOGGER = Logger.getLogger(JwtFilter.class.getName());

    @Inject
    OidcConfig oidcConfig;

    private JWTVerifier jwtVerifier;

> 

Reviewer: ```suggestion

```

---
layout: blog_post

> title: "Use Jakarta EE 10 with OpenID Connect Authentication"

Reviewer: This scores a 57 on our [title analyzer](https://blog-title-analyzer.oktadev.dev/). 

Other options:

- Use Jakarta EE 10 with OIDC Authentication: 58
- Jakarta EE 10 with OpenID Connect Authentication: 59
- Get Started with Jakarta EE 10 and OpenID Connect Authentication: 55

Reviewer: Keyword traffic:

- jakarta ee: 9900
- jakarta security: 170
- jakarta ee 10: 8080

To me, it seems we should optimize for "Jakarta EE" and "Jakarta EE 10" in the title and first few headings.

author: andrew-hughes
by: contractor
communities: [java,security]
description: "Use Jakarta EE 10 to build a secure Java web application using OpenID Connect and Auth0."
tags: [java, jakartaee]
image:
type: conversion
---

Jakarta EE 10 includes a new authentication mechanism: OpenID Connect! This can be added to a Jakarta EE servlet using the new `@OpenIdAuthenticationMechanismDefinition` annotation. 

In this tutorial, you are going to see how to implement a web application with OpenID Connect (OIDC) authentication and use Auth0 as the OIDC provider. You are also going to see one way to secure an API and authenticate using JSON Web Tokens (JWTs). This will all be accomplished using WildFly as the Jakarta EE runtime.

This stack includes a lot of technologies. I'm going to introduce them briefly below. If you're comfortable with all those terms and just want to get to the code, **feel free to skip ahead to the prerequisites section**.

{% include toc.md %}

> ## Jakarta vs Java, EE vs SE

Reviewer: ```suggestion
## Jakarta EE vs Java EE
```

Jakarta EE is Jakarta Enterprise Edition, formerly known as Java EE. The name and framework packages were migrated when Oracle gave Java EE to the Eclipse Foundation because Oracle still has the rights to the Java brand and did not open-source absolutely everything that was in the `javax.*` namespace. Thus, Jakarta EE is the Eclipse-owned and now totally open-source Java EE (You may have recently had to change some packages from `javax` to `jakarta`. This is why.)

Enterprise Edition is built on top of Jakarta (that is, Java) SE, or Standard Edition. Jakarta SE is the more lightweight Java version that provides a basic cross-platform runtime. Enterprise Edition is assumed to be running on an application server and adds libraries intended for larger-scale, multi-user applications. 

> To run an SE application, all you need is the JRE (Java runtime environment) for a compatible version of Java. Enterprise Edition, however, requires a more complete runtime environment and has a lot more possible modules and configuration options. To see a list of Jakarta EE compatible products, you can look at [the Jakarta website](https://jakarta.ee/compatibility/). A few examples are Open Liberty, Payara, WildFly, GlassFish, and TomEE.

Reviewer: ```suggestion
### What about Java SE? 

To run a Java SE application, all you need is the JRE (Java runtime environment) for a compatible version of Java. Enterprise Edition, however, requires a more complete runtime environment and has a lot more possible modules and configuration options. To see a list of Jakarta EE compatible products, you can look at [the Jakarta website](https://jakarta.ee/compatibility/). A few examples are Open Liberty, Payara, WildFly, GlassFish, and TomEE.
```

As of the time I wrote this tutorial, Jakarta EE 10 was a very new release, and only three frameworks supported version 10: Eclipse GlassFish, Payara Server Community, and WildFly. 

> ### WildFly

Reviewer: ```suggestion
### WildFly for Jakarta EE 10
```

I chose to use [WildFly](https://www.wildfly.org/) as my Jakarta EE runtime. Hantsy Bai created a great example project that was a big help. Check out [the GitHub project repository page](https://github.com/hantsy/jakartaee10-sandbox). Thanks, Hantsy Bai! Super helpful.

WildFly is an open-source community project sponsored by Red Hat. It bills itself as a "flexible, lightweight, managed application runtime" that is "based on Jakarta EE and provides rich enterprise capabilities in easy to consume frameworks that eliminate boilerplate and reduce technical burden." It is a modular, standards-based runtime for Jakarta EE applications.

### Jakarta Security 3.0

> The exciting thing about Jakarta 10 (from a security perspective) is that it includes a new OIDC implementation in the Security 3.0 specification. OpenID Connect is an authentication protocol. This protocol is implemented by many third-party vendors, such as Auth0 and Okta, making it relatively easy to add secure login to an application. Jakarta EE 10 Security 3.0 provides an annotation-based configuration to add OIDC authentication to servlets.

Reviewer: ```suggestion
The exciting thing about Jakarta EE 10 (from a security perspective) is that it includes a new OIDC implementation in the Security 3.0 specification. OpenID Connect is an authentication protocol. This protocol is implemented by many third-party vendors, such as Auth0 and Okta, making it relatively easy to add secure login to an application. Jakarta EE 10 Security 3.0 provides an annotation-based configuration to add OIDC authentication to servlets.
```

You can check out [the docs for Jakarta EE 10 Security 3.0 OIDC here](https://jakarta.ee/specifications/security/3.0/jakarta-security-spec-3.0.html#openid-connect-annotation).

> ## Prerequisites

Reviewer: I decided to make this a heading because there's a lot of intro before you get to this section. I could also see rolling it into the next section.

Before you start, please make sure you have the following prerequisites installed (or install them now).

- [Java 17](https://adoptium.net/): or use [SDKMAN!](https://sdkman.io/) to manage and install multiple versions (the Jakarta EE spec says 11 and up is supported, but I wrote this tutorial assuming version 17)
- [Auth0 CLI](https://github.com/auth0/auth0-cli#installation): the Auth0 command-line interface
- [HTTPie](https://httpie.org/doc#installation): a simple tool for making HTTP requests from a Bash shell

**You will need a free Auth0 developer account** if you don't already have one. Go ahead and sign up for an Auth0 account using [their sign-up page](https://auth0.com/signup).

Clone this tutorial's example from [its GitHub repository](https://github.com/oktadev/auth0-jakarta-ee-oidc-example).

```shell

> git clone https://github.com/oktadev/auth0-jakarta-ee-oidc-example.git

Reviewer: @moksamedia Please let me know the repo for this example and I can fork it to `@oktadev/auth0-jakarta-ee-oidc-example`.

I won't reproduce the entire `pom.xml` file here, but I want to point out a few things.

First, take a look at the dependencies. The only dependency required for Jakarta EE is the first one (`jakarta.jakartaee-api`). 

```xml
<dependencies>
    <dependency>
        <groupId>jakarta.platform</groupId>
        <artifactId>jakarta.jakartaee-api</artifactId>
        <version>${jakartaee-api.version}</version>
        <scope>provided</scope>
    </dependency>
</dependencies>
```

The WildFly plugin is included and configured in the block below.

```xml
<plugin>
    <groupId>org.wildfly.plugins</groupId>
    <artifactId>wildfly-maven-plugin</artifactId>
    <version>${wildfly-maven-plugin.version}</version>
    <configuration>
        <commands>
            <command>/subsystem=undertow/application-security-domain=other:write-attribute(name=integrated-jaspi, value=false)</command>
            <command>reload</command>
        </commands>
    </configuration>
</plugin>
```

The docs for [the WildFly Maven plugin are here](https://docs.wildfly.org/wildfly-maven-plugin/). Except for the cryptic `<command></command>` block, the plugin is pretty simple and easy to use. 

> It took a little digging to figure out, but the obscure command block *is* required, at least according to the experts I asked. It disables integrated JASPI in the server and instead delegates validation of credentials to a non-integrated `ServerAuthModule`. This allows for identities to be dynamically created instead of statically stored in an integrated security domain. Take a look at the [Elytron and Java EE Security section of the docs](https://docs.wildfly.org/26/WildFly_Elytron_Security.html#Elytron_and_Java_EE_Security) for more on this.

Reviewer: ```suggestion
It took a little digging to figure out, but the obscure command block *is* required, at least according to the experts I asked. It disables integrated JASPI (Java Authentication SPI for Containers) in the server and instead delegates validation of credentials to a non-integrated `ServerAuthModule`. This allows for identities to be dynamically created instead of statically stored in an integrated security domain. Take a look at the [Elytron and Java EE Security section of the docs](https://docs.wildfly.org/26/WildFly_Elytron_Security.html#Elytron_and_Java_EE_Security) for more on this.
```

## Project structure and configuration

The files in the `src` directory are listed below. There are three different services: 

1. the OIDC-protected servlet;
2. the API servlet protected by a JWT authentication filter; and
3. an unprotected servlet.

```text

> src
> └── main
>     ├── java
>     │   └── com
>     │       └── demo
>     │           ├── CallbackServlet.java  // OIDC callback handler
>     │           ├── ProtectedServlet.java  // OIDC-handling servlet endpoint
>     │           ├── PublicServlet.java  // Public endpoint
>     │           ├── JwtFilter.java  // Verifies JWT and secures ApiServlet
>     │           ├── ApiServlet.java // API protected by filter
>     │           └── OpenIdConfig.java  // Loads openid.properties
>     ├── resources
>     │   ├── logging.properties  // Simple console logging configuration
>     │   ├── META-INF
>     │   │   ├── beans.xml  // Declare some provided dependencies for deployment
>     │   │   └── MANIFEST.MF  // Configure CDI (Contexts and Dependency Injection)
>     │   └── openid.properties  // OpenID config properties
>     └── webapp
>         └── WEB-INF
>             └── jboss-web.xml  // Configures context root to '/'

Reviewer: ```suggestion
src
└── main
    ├── java
    │   └── com
    │       └── demo
    │           ├── ApiServlet.java // API protected by filter
    │           ├── CallbackServlet.java  // OIDC callback handler
    │           ├── JwtFilter.java  // Verifies JWT and secures ApiServlet
    │           ├── OidcConfig.java  // Loads oidc.properties
    │           ├── ProtectedServlet.java  // OIDC-handling servlet endpoint
    │           └── PublicServlet.java  // Public endpoint
    ├── resources
    │   ├── META-INF
    │   │   ├── beans.xml  // Declare some provided dependencies for deployment
    │   │   └── MANIFEST.MF  // Configure CDI (Contexts and Dependency Injection)
    │   ├── logging.properties  // Simple console logging configuration
    │   └── oidc.properties  // OpenID Connect config properties
    └── webapp
        └── WEB-INF
            └── jboss-web.xml  // Configures context root to '/'
```

Reorder to match IntelliJ and rename so there's no confusion between OpenID and OpenID Connect.

```

> When the application loads, the OpenID properties are loaded from `openid.properties` by the `OpenIDConfig` class. These values are used by the `JwtFilter` to create the class that verifies JSON Web Tokens. These properties are also used by the `ProtectedServlet` in the `@OpenIdAuthenticationMechanismDefinition` annotation to configure OIDC.

Reviewer: ```suggestion
When the application loads, the OpenID Connect properties are loaded from `oidc.properties` by the `OidcConfig` class. These values are used by the `JwtFilter` to create the class that verifies JSON Web Tokens. These properties are also used by the `ProtectedServlet` in the `@OpenIdAuthenticationMechanismDefinition` annotation to configure OIDC.
```

## Create an Auth0 OIDC application

If you have not already, install the [Auth0 CLI](https://github.com/auth0/auth0-cli) and run `auth0 login` in a terminal. As I write this, the Auth0 CLI 1.0 version is in beta. It adds some new features that I'll mention. You can take a look at [the release here](https://github.com/auth0/auth0-cli/releases/tag/v1.0.0-beta.1).

```bash
Waiting for the login to complete in the browser... done

 ▸    Successfully logged in.
 ▸    Tenant: dev-0xb84jzp.us.auth0.com
```

Take note of the domain listed as the tenet. This is your Auth0 domain. If you need to find it again later, you can use `auth0 tenants list`.

Now use the Auth0 CLI to create an OpenID Connect (OIDC) application.  From the project base directory, run the following.

```bash
auth0 apps create
```

Use the following values:

> - **Name**: `javartaee-demo`
> 
> - **Description**: whatever you like, or leave blank

Reviewer: ```suggestion
- **Name**: `jakartaee-demo`
- **Description**: whatever you like, or leave blank
```

- **Type**: `Regular Web Application`
- **Callback URLs**: `http://localhost:8080/callback`
- **Allowed Logout URLs**: `http://localhost:8080`

> The console output shows you the Auth0 domain and the OIDC client ID. However, you also need the client secret. On the 1.0 version of the Auth0 CLI you can show the client sercret with the `--reveal-secrets` command. However, for previous versions, you have to get the client secret by logging into Auth0. Type the following:

Reviewer: ```suggestion
The console output shows you the Auth0 domain and the OIDC client ID. However, you also need the client secret. On the 1.0 version of the Auth0 CLI you can show the client secret by adding `--reveal-secrets` to the `apps create` command. However, for previous versions, you have to get the client secret by logging into Auth0. Type the following:
```

```bash
auth0 apps open
```

Select the OIDC app (or client) you just created from the list. This will open the OIDC application on the Auth0 dashboard.

{% img blog/jakartaee-auth0/oidc-application-auth0.png alt:"Auth0 OIDC App" width:"600" %}{: .center-image }

> Fill in the three values in `src/main/resources/openid.properties`. Replace the bracketed values with the values from the OIDC application page on the Auth0 dashboard.

Reviewer: ```suggestion
Fill in the three values in `src/main/resources/oidc.properties`. Replace the bracketed values with the values from the OIDC application page on the Auth0 dashboard.
```

```properties

> issuerUri=<your-auth0-domain>

Reviewer: ```suggestion
issuerUri=https://<your-auth0-domain>
```

Change the code for the action to the following.

```js
exports.onExecutePostLogin = async (event, api) => {
  const namespace = 'http://www.jakartaee.demo';
  if (event.authorization) {
    api.idToken.setCustomClaim('preferred_username', event.user.email);
    api.idToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
    api.accessToken.setCustomClaim(`${namespace}/roles`, event.authorization.roles);
  }
}
```

Click on **Deploy**. 

Click on the **Add to flow** link in the popup window that slides in (if you miss this, you can find the new action under the custom action tab back in the flow panel).

Drag the **Add Roles** action over under the **Rules (legacy)** action. 

{% img blog/jakartaee-auth0/auth0-action-flow.png alt:"Auth0 Action Flow" width:"400" %}{: .center-image }

Click **Apply** (top right of the panel).

## Explore the `ProtectedServlet` and OIDC flow

Let's look at the `ProtectedServlet` first. This is the class that defines the OIDC annotation and will redirect to Auth0 to handle OIDC authentication.

`src/main/java/com/demo/ProtectedServlet.java`

```java
package com.demo;

...

> @OpenIdAuthenticationMechanismDefinition(
>         providerURI = "${openIdConfig.issuerUri}",
>         clientId = "${openIdConfig.clientId}",
>         clientSecret = "${openIdConfig.clientSecret}",
>         redirectURI = "${baseURL}/callback",
>         // default 500ms caused timeouts for me
>         jwksConnectTimeout = 5000,
>         jwksReadTimeout = 5000,
>         // Auth0 requires the audience to be set to the default API
>         extraParameters = {"audience=https://${openIdConfig.issuerUri}/api/v2/"},
>         // read the roles from Auth0 custom claim
>         claimsDefinition = @ClaimsDefinition(callerGroupsClaim = "http://www.jakartaee.demo/roles")
> )

Reviewer: ```suggestion
@OpenIdAuthenticationMechanismDefinition(
    providerURI = "${openIdConfig.issuerUri}",
    clientId = "${openIdConfig.clientId}",
    clientSecret = "${openIdConfig.clientSecret}",
    redirectURI = "${baseURL}/callback",
    // default 500ms caused timeouts for me
    jwksConnectTimeout = 5000,
    jwksReadTimeout = 5000,
    extraParameters = {"audience=https://${openIdConfig.issuerUri}/api/v2/"},
    claimsDefinition = @ClaimsDefinition(callerGroupsClaim = "http://www.jakartaee.demo/roles")
)
```

@WebServlet("/protected")
@ServletSecurity(

>         @HttpConstraint(rolesAllowed = "Everyone")

Reviewer: ```suggestion
    @HttpConstraint(rolesAllowed = "Everyone")
```

At this point, the user is successfully authenticated. If you look at the callback servlet (shown below), you'll see that it simply redirects the user back to the `/protected` servlet. 

```java
package com.demo;

...

@WebServlet("/callback")
public class CallbackServlet extends HttpServlet {

    private static final Logger LOGGER = Logger.getLogger(CallbackServlet.class.getName());

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        String referer = (String) request.getSession().getAttribute("Referer");
        String redirectTo = referer != null ? referer : request.getContextPath() + "/protected";
        LOGGER.info("OIDC callback success. Redirecting to: " + redirectTo);
        response.sendRedirect(redirectTo);

>     }
> 
> }
> 

Reviewer: ```suggestion
    }
}
```

- Auth0 accepts the authorization code, verifies it, and returns an access token (and possibly an identity token) to the Jakarta EE Security 3.0 framework.
- Once a verified JWT is received and unpacked, the user is authenticated and the `@ServletSecurity` annotation requirement is checked. If the user is a member of the `Everyone` group, the `ProtectedServlet.doGet()` method is called.
- The  `ProtectedServlet.doGet()` method programmatically redirects back to `/protected`.

All of that happened above when you logged into Auth0 and loaded the protected servlet. Since this servlet handily prints out the JWT, I thought it would be nice to see how to secure a web API using a JWT, which is what you'll see in the next section.

## Use a JWT to access the protected API

The `ApiServlet` file defines an API servlet.

`src/main/java/com/demo/ApiServlet.java`

```java
package com.demo;

...

@WebServlet("/api/protected")
public class ApiServlet extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {

        DecodedJWT jwt = (DecodedJWT)request.getAttribute("jwt");

>         response.setContentType("text");

Reviewer: ```suggestion
        DecodedJWT jwt = (DecodedJWT) request.getAttribute("jwt");
        response.setContentType("text/plain");
```

        response.getWriter().println("Welcome, " + jwt.getClaims().get("sub"));
        response.getWriter().println(jwt.getClaims());
    }

> }
> 
> ```

Reviewer: ```suggestion
}
```
```

This servlet by itself is not at all secure and would be public without the `JwtFilter` class, which is shown below. The filter intercepts any requests matching the `/api/*` URL pattern and denies them if they do not have a valid JWT.

`src/main/java/com/demo/JwtFilter.java`

```java
package com.demo;

...

@WebFilter(filterName = "jwtFilter", urlPatterns = "/api/*")
public class JwtFilter implements Filter {

    private static final Logger LOGGER = Logger.getLogger(JwtFilter.class.getName());

    @Inject

>     OpenIdConfig openIdConfig;

Reviewer: ```suggestion
    OidcConfig oidcConfig;
```

    private JWTVerifier jwtVerifier;

    @Override
    public void init(FilterConfig filterConfig) {
        LOGGER.info("Auth0 jwtVerifier initialized for issuer:" + openIdConfig.getIssuerUri());
    }

    @Override
    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse,
                         FilterChain chain) throws IOException, ServletException {

        HttpServletRequest request = (HttpServletRequest) servletRequest;
        HttpServletResponse response = (HttpServletResponse) servletResponse;

        LOGGER.info("In JwtFilter, path: " + request.getRequestURI());

        // Get access token from authorization header
        String authHeader = request.getHeader("authorization");
        if (authHeader == null) {
            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
            response.getOutputStream().print("Unauthorized");
            return;
        } else {
            String accessToken = authHeader.substring(authHeader.indexOf("Bearer ") + 7);
            LOGGER.info("accesstoken: " + request.getRequestURI());

>             JwkProvider provider = new UrlJwkProvider(openIdConfig.getIssuerUri());

Reviewer: ```suggestion
            JwkProvider provider = new UrlJwkProvider(oidcConfig.getIssuerUri());
```

            try {
                DecodedJWT jwt = JWT.decode(accessToken);
                // Get the kid from received JWT token
                Jwk jwk = provider.get(jwt.getKeyId());

                Algorithm algorithm = Algorithm.RSA256((RSAPublicKey) jwk.getPublicKey(), null);

>                 JWTVerifier verifier = JWT.require(algorithm)
>                         .withIssuer(openIdConfig.getIssuerUri())
>                         .build();

Reviewer: ```suggestion
                JWTVerifier verifier = JWT.require(algorithm)
                    .withIssuer(oidcConfig.getIssuerUri())
                    .build();
```

                jwt = verifier.verify(accessToken);
                LOGGER.info("JWT decoded. sub=" + jwt.getClaims().get("sub"));
                request.setAttribute("jwt", jwt);

            } 
            
            ...

        }

        chain.doFilter(request, response);
    }

    @Override
    public void destroy() {
    }

}
```

This code uses Auth0's JWT verifier for Java. Auth0 [has good docs on JWT verification](https://auth0.com/docs/secure/tokens/json-web-tokens/validate-json-web-tokens). If a valid JWT is found and decoded, it is saved in a request attribute and the request is allowed to continue.

Give it a try. Start the project.

```bash
./mvnw wildfly:run
```

Make a request to the protected API endpoint (not the OIDC endpoint).

```bash
http :8080/api/protected
```

You'll get:

> ```Bash

Reviewer: ```suggestion
```bash
```

Date: Tue, 27 Jan 2023 15:02:59 GMT

Welcome, andrew.hughes@mail.com
...

```

## Keep learning with Jakarta EE and Auth0

You just built a Jakarta Enterprise Edition application that used the new OpenID Connect annotation and implementation built into Jakarta EE 10. You used Auth0 as the OIDC and OAuth 2.0 provider, and you saw how to implement both SSO and JWT authentication.

You can find the source code for this example on GitHub in the [@oktadev/okta-jakartaee-oidc-example](https://github.com/oktadev/okta-jakartaee-oidc-example) repository.

If you liked this post, there's a good chance you'll like similar ones:

- [Introducing Spring Native for JHipster: Serverless Full-Stack Made Easy](/blog/2022/03/03/spring-native-jhipster)
- [Add Authentication to Your Vanilla JavaScript App in 20 Minutes](/blog/2018/06/05/authentication-vanilla-js)
- [Mobile Development with Ionic, React Native, and JHipster](/blog/2020/04/27/mobile-development-ionic-react-native-jhipster)
- [Fast Java Made Easy with Quarkus and JHipster](/blog/2021/03/08/jhipster-quarkus-oidc)
- [Build a CRUD App with Vue.js, Spring Boot, and Kotlin](/blog/2020/06/26/spring-boot-vue-kotlin)

> - [Add OpenID Connect to Angular Apps Quickly](/blog/2022/02/11/angular-auth0-quickly)

Reviewer: TODO: Update this list to use more recent posts.

Reviewer: @dan-auth0 Do you have a list of popular/related Java blog posts we should add here in the CTA (call to action)?

    }

}
```

This code uses Auth0's JWT verifier for Java. Auth0 [has good docs on JWT verification](https://auth0.com/docs/secure/tokens/json-web-tokens/validate-json-web-tokens). If a valid JWT is found and decoded, it is saved in a request attribute, and the request is allowed to continue.

Give it a try. Start the project.

```bash
./mvnw wildfly:run
```

Make a request to the protected API endpoint (not the OIDC endpoint).

```bash
http :8080/api/protected
```

You'll get:

```bash
HTTP/1.1 401 Unauthorized
```

Now, use your OIDC endpoint to retrieve a token. Using a browser, open `http://localhost:8080/protected`.

Authenticate with Auth0. When you are redirected back to the protected servlet page, copy the token value and save it as a variable in a new shell.

```bash
TOKEN=eyJraWQiOiJqY3dpbGpUcGVZSG1Jajl6ODR3LV...
```

In that same shell, make a request to the protected API endpoint using the token.

```bash
http :8080/api/protected "Authorization: Bearer $TOKEN"
```

It should return something like the following.

```bash
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 401
Content-Type: text;charset=ISO-8859-1
Date: Tue, 23 Feb 2023 07:48:59 GMT

> Welcome, andrew.hughes@mail.com

Reviewer: @moksamedia This results in:

```
Welcome, "auth0|61bcbc76f64d4a0072af8a1d"
{sub="auth0|61bcbc76f64d4a0072af8a1d", aud=["https://dev-06bzs1cu.us.auth0.com/api/v2/","https://dev-06bzs1cu.us.auth0.com/userinfo"], azp="YwvfJ8soOxh3g1kmehMr1pXx6GgBIqMh", scope="openid profile email", iss="https://dev-06bzs1cu.us.auth0.com/", http://www.jakartaee.demo/roles=["Everyone","ROLE_ADMIN","ROLE_USER"], exp=1677268045, iat=1677181645}
```

Can you fix the code to display the user's email or update the post to reflect the proper `sub`?

Author: I updated to ApiFilter to pull the ID token from the userinfo endpoint and use that in the ApiServlet to show the email address.

Reviewer: Thanks, good stuff!

}
```

> This code uses Auth0's JWT verifier for Java. Auth0 [has good docs on JWT verification](https://auth0.com/docs/secure/tokens/json-web-tokens/validate-json-web-tokens). If a valid JWT is found and decoded, it is saved in a request attribute. The access token, however, has very limited user information in it (just the user ID as the `sub` claim). The filter code uses the access token to request the ID token from the `userinfo` endpoint. This returns more complete user information, such as the email address and preferred name. This is deserialized into the `IdToken` Java object and stored in another request attribute. These are read in the `ApiServlet`.

Reviewer: ```suggestion
This code uses Auth0's JWT verifier for Java. Auth0 [has good docs on JWT validation](https://auth0.com/docs/secure/tokens/json-web-tokens/validate-json-web-tokens). If a valid JWT is found and decoded, it is saved in a request attribute. The access token, however, has minimal user information in it (just the user ID as the `sub` claim). The filter code uses the access token to request the ID token from the `userinfo` endpoint. This returns more complete user information, such as the email address and preferred name. This is deserialized into the `IdToken` Java object and stored in another request attribute. These are read in the `ApiServlet`.
```
